ovn-tester/ovn_workload.py

```diff
@@ -350,206 +367,218 @@ def ping_ports(self, cluster, ports):
             'allow-related'
         )
-        self.nbctl.acl_add(
+        await self.nbctl.acl_add(
             self.pg.name, 'to-lport', ACL_NETPOL_ALLOW_PRIO, 'port-group',
             f'ip4.dst == \\${self.sub_as[dst].name} && '
             f'inport == @{self.sub_pg[src].name}',
             'allow-related'
         )
 
     @ovn_stats.timeit
-    def allow_from_external(self, external_ips, include_ext_gw=False):
-        self.enforce()
+    async def allow_from_external(self, external_ips, include_ext_gw=False):
+        await self.enforce()
         # If requested, include the ext-gw of the first port in the namespace
         # so we can check that this rule is enforced.
         if include_ext_gw:
             assert(len(self.ports) > 0)
             external_ips.append(self.ports[0].ext_gw)
         ips = [str(ip) for ip in external_ips]
-        self.nbctl.acl_add(
+        await self.nbctl.acl_add(
             self.pg.name, 'to-lport', ACL_NETPOL_ALLOW_PRIO, 'port-group',
             f'ip4.src == {{{",".join(ips)}}} && outport == @{self.pg.name}',
             'allow-related'
         )
 
     @ovn_stats.timeit
-    def check_enforcing_internal(self):
+    async def check_enforcing_internal(self):
         # "Random" check that first pod can reach last pod in the namespace.
         if len(self.ports) > 1:
             src = self.ports[0]
             dst = self.ports[-1]
             worker = src.metadata
-            worker.ping_port(self.cluster, src, dst.ip)
+            await worker.ping_port(self.cluster, src, dst.ip)
 
     @ovn_stats.timeit
-    def check_enforcing_external(self):
+    async def check_enforcing_external(self):
         if len(self.ports) > 0:
             dst = self.ports[0]
             worker = dst.metadata
-            worker.ping_external(self.cluster, dst)
+            await worker.ping_external(self.cluster, dst)
 
     @ovn_stats.timeit
-    def check_enforcing_cross_ns(self, ns):
+    async def check_enforcing_cross_ns(self, ns):
         if len(self.ports) > 0 and len(ns.ports) > 0:
             dst = ns.ports[0]
             src = self.ports[0]
             worker = src.metadata
-            worker.ping_port(self.cluster, src, dst.ip)
+            await worker.ping_port(self.cluster, src, dst.ip)
 
-    def create_load_balancer(self):
-        self.load_balancer = lb.OvnLoadBalancer(f'lb_{self.name}', self.nbctl)
+    async def create_load_balancer(self):
+        self.load_balancer = await create_load_balancer(f'lb_{self.name}',
+                                                        self.nbctl)
 
     @ovn_stats.timeit
-    def provision_vips_to_load_balancers(self, backend_lists):
-        vip_net = DEFAULT_NS_VIP_SUBNET.next(self.cluster.n_ns)
-        n_vips = len(self.load_balancer.vips.keys())
-        vip_ip = vip_net.ip.__add__(n_vips + 1)
+    async def provision_vips_to_load_balancers(self, backend_lists):
+        async with self.lock:
```

Comment:
I think we can avoid this lock all together if we split the `add_vips` API into, e.g., `reserve_vips()` (synchronous and just adding the items to the load balancer vips dict and returning them) and `add_vips()` coroutine to actually set the vips.  If you agree, I'll let it up to you whether we should make the change in this PR or in a separate one.

Reply:
I was messing with this just now, and removing the locking is going to be non-trivial. 

Adding VIPs is not an additive operation. Under the hood it calls `ovn-nbctl set Load_Balancer vips ...`. That means each individual task is re-setting the entire VIP list instead of just adding their own VIPs to the configured load balancer. This means that each task needs to have exclusive access to the load balancer.

The VIPs are added in batches, meaning that each time we add new VIPs, we might end up making multiple ovn-nbctl calls per load balancer. Since all of those ovn-nbctl calls are asynchronous, it gives other tasks the opportunity to foul up the configured VIPs. So even if we could make the locking finer-grained, I think we'd still need locking as long as we're batching the VIPs.

Having said that, I actually noticed that there's a bug in the current `add_vips()` method due to the batching code. Since we are setting the VIPs on the load balancer instead of adding to them, it means that each batch undoes the previous batch's operation. Effectively it means that a load balancer will never have more than `MAX_VIPS_IN_BATCH` (currently 500) vips configured. This may not be as big an issue with load balancer per service, but it's still a bug. Since the operation is not additive, I actually have no idea how batching can work here.

If we could make adding VIPs additive, then this would be easy since each task would be performing independent non-overlapping operations. It would also allow for proper batching.

Reply:
> I was messing with this just now, and removing the locking is going to be non-trivial.
> 
> Adding VIPs is not an additive operation. Under the hood it calls `ovn-nbctl set Load_Balancer vips ...`. That means each individual task is re-setting the entire VIP list instead of just adding their own VIPs to the configured load balancer. This means that each task needs to have exclusive access to the load balancer.

Actually, I think adding VIPs *is* an additive operation, it translates to something equivalent to:

```
$ ovn-nbctl lb-add lb-test 42.42.42.42:4242 21.21.21.21:2121
$ ovn-nbctl list load_balancer
_uuid               : 00874eb0-2266-43da-870c-3e8f9cb55617
name                : lb-test
vips                : {"42.42.42.42:4242"="21.21.21.21:2121"}
$ ovn-nbctl set load_balancer lb-test vips:\"43.43.43.43:4343\"=\"22.22.22.22:222\"
$ ovn-nbctl list load_balancer
_uuid               : 00874eb0-2266-43da-870c-3e8f9cb55617
name                : lb-test
vips                : {"42.42.42.42:4242"="21.21.21.21:2121", "43.43.43.43:4343"="22.22.22.22:222"
```

> 
> The VIPs are added in batches, meaning that each time we add new VIPs, we might end up making multiple ovn-nbctl calls per load balancer. Since all of those ovn-nbctl calls are asynchronous, it gives other tasks the opportunity to foul up the configured VIPs. So even if we could make the locking finer-grained, I think we'd still need locking as long as we're batching the VIPs.
> 
> Having said that, I actually noticed that there's a bug in the current `add_vips()` method due to the batching code. Since we are setting the VIPs on the load balancer instead of adding to them, it means that each batch undoes the previous batch's operation. Effectively it means that a load balancer will never have more than `MAX_VIPS_IN_BATCH` (currently 500) vips configured. This may not be as big an issue with load balancer per service, but it's still a bug. Since the operation is not additive, I actually have no idea how batching can work here.

As mentioned above, unless I'm missing something, the operation is already additive.

> 
> If we could make adding VIPs additive, then this would be easy since each task would be performing independent non-overlapping operations. It would also allow for proper batching.

Reply:
> Actually, I think adding VIPs **is** an additive operation, it translates to something equivalent to:
> 
> ```
> $ ovn-nbctl lb-add lb-test 42.42.42.42:4242 21.21.21.21:2121
> $ ovn-nbctl list load_balancer
> _uuid               : 00874eb0-2266-43da-870c-3e8f9cb55617
> name                : lb-test
> vips                : {"42.42.42.42:4242"="21.21.21.21:2121"}
> $ ovn-nbctl set load_balancer lb-test vips:\"43.43.43.43:4343\"=\"22.22.22.22:222\"
> $ ovn-nbctl list load_balancer
> _uuid               : 00874eb0-2266-43da-870c-3e8f9cb55617
> name                : lb-test
> vips                : {"42.42.42.42:4242"="21.21.21.21:2121", "43.43.43.43:4343"="22.22.22.22:222"
> ```

That works as long as the key for the VIP is different each time. It appears our tests do not attempt to add backends to existing VIPs, so you're right this should work for our purposes.

If we can guarantee that the set of keys in the dictionary of VIPs passed to `add_vips()` does not intersect with the set of keys already configured on the load balancer, we could also simplify some of the code in `add_vips()`

Reply:
Shall we raise an exception then if a test tries to add backends to an already configured VIP?